### PR TITLE
Add socket-based kubelet pod resources API.

### PR DESCRIPTION
**Pull request by dashpole**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This introduces a new kubelet socket, which provides read-only information about pods, and serves a grpc service.  The initial use-case for this is to support out-of-tree monitoring agents, which need to be able to correlate devices and containers.

**Which issue(s) this PR fixes**:
Feature Issue: https://github.com/kubernetes/enhancements/issues/606
KEP: https://github.com/kubernetes/community/pull/2454

**Special notes for your reviewer**:
The API is in `pkg/kubelet/apis/podresources/v1alpha1`.
This adds to the device plugin e2e_node test to test its functionality.

```release-note
Add a kubelet socket which serves an grpc service containing the devices used by containers on the node.
```

cc @RenaudWasTaken @jiayingz @vikaschoudhary16 @vishh @derekwaynecarr 
cc @kubernetes/sig-node-api-reviews

### CODE REVIEW

**Comment by dashpole** [comment]
cc @guptaNswati

**Comment by jeremyeder** [comment]
@vikaschoudhary16 can you take a look?

File: hack/update-generated-pod-resources-dockerized.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
```
Comment (RenaudWasTaken) [inline hack/update-generated-pod-resources-dockerized.sh]: You should move the files in the hack directory in a separate commit

File: pkg/kubelet/apis/podresources/client.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+	"google.golang.org/grpc"
+
+	podresourcesapi "k8s.io/kubernetes/pkg/kubelet/apis/podresources/v1alpha1"
+	"k8s.io/kubernetes/pkg/kubelet/util"
+)
+
+// GetClient returns a client for the PodResourcesLister grpc service
+func GetClient(socket string, connectionTimeout time.Duration, maxMsgSize int) (podresourcesapi.PodResourcesListerClient, *grpc.ClientConn, error) {
+	addr, dailer, err := util.GetAddressAndDialer(socket)
```
Comment (RenaudWasTaken) [inline pkg/kubelet/apis/podresources/client.go]: nit: `dialer` not `dailer`

File: pkg/kubelet/apis/podresources/server_test.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+		{
+			desc:             "no pods",
+			pods:             []*v1.Pod{},
+			devices:          []*v1alpha1.ContainerDevices{},
+			expectedResponse: &v1alpha1.ListPodResourcesResponse{},
+		},
+		{
+			desc: "pod without devices",
+			pods: []*v1.Pod{
+				{
```
Comment (RenaudWasTaken) [inline pkg/kubelet/apis/podresources/server_test.go]: Do you think a helper would make more sense here?

File: pkg/kubelet/apis/podresources/v1alpha1/api.proto [hunk 4]
```diff
@@ -0,0 +1,43 @@
+// To regenerate api.pb.go run hack/update-generated-pod-resources.sh
```
Comment (RenaudWasTaken) [inline pkg/kubelet/apis/podresources/v1alpha1/api.proto]: Split in two or three different commits:
- gRPC API and constants
- [Optional] the server and the tests
- Generated golang code

File: pkg/kubelet/apis/podresources/v1alpha1/api.proto [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+service PodResourcesLister {
+    rpc List(ListPodResourcesRequest) returns (ListPodResourcesResponse) {}
+}
+ // ListPodResourcesRequest is the request made to the PodResourcesLister service
+message ListPodResourcesRequest {}
+ // ListPodResourcesResponse is the response returned by List function
+message ListPodResourcesResponse {
+    repeated PodResources pod_resources = 1;
+}
+ // PodResources contains information about the node resources assigned to a pod
```
Comment (RenaudWasTaken) [inline pkg/kubelet/apis/podresources/v1alpha1/api.proto]: Is there a space before your comment? In that case, nit: Remove the space :)

File: pkg/kubelet/kubelet.go [hunk 6]
```diff
@@ -1256,6 +1257,9 @@ func (kl *Kubelet) setupDataDirs() error {
 	if err := os.MkdirAll(kl.getPluginsDir(), 0750); err != nil {
 		return fmt.Errorf("error creating plugins directory: %v", err)
 	}
+	if err := os.MkdirAll(kl.getPodResourcesDir(), 0750); err != nil {
```
Comment (RenaudWasTaken) [inline pkg/kubelet/kubelet.go]: This should probably part of an `Init` function in the server rather than being at the top level

File: cmd/kubelet/app/server.go [hunk 7]
```diff
@@ -1015,6 +1015,8 @@ func startKubelet(k kubelet.Bootstrap, podCfg *config.PodConfig, kubeCfg *kubele
 	if kubeCfg.ReadOnlyPort > 0 {
 		go k.ListenAndServeReadOnly(net.ParseIP(kubeCfg.Address), uint(kubeCfg.ReadOnlyPort))
 	}
+
+	go k.ListenAndServePodResources()
```
Comment (RenaudWasTaken) [inline cmd/kubelet/app/server.go]: Should this be feature gated?

**Review by RenaudWasTaken — COMMENTED** [review]
Overall the implementation looks good.

**Comment by AishSundar** [comment]
/milestone v1.13 since this is related to https://github.com/kubernetes/enhancements/issues/606

File: hack/update-generated-pod-resources-dockerized.sh (hunk 1 above)
Comment (dashpole) [reply hack/update-generated-pod-resources-dockerized.sh]: Added hack and generated files into a separate commit.

File: pkg/kubelet/apis/podresources/client.go (hunk 2 above)
Comment (dashpole) [reply pkg/kubelet/apis/podresources/client.go]: done

File: pkg/kubelet/apis/podresources/server_test.go (hunk 3 above)
Comment (dashpole) [reply pkg/kubelet/apis/podresources/server_test.go]: We only make 2 pods, so I don't think its worth it.  if/when we add to these, it may make more sense to add a helper.

File: pkg/kubelet/apis/podresources/v1alpha1/api.proto (hunk 4 above)
Comment (dashpole) [reply pkg/kubelet/apis/podresources/v1alpha1/api.proto]: done.  I decided to do:

* grpc API
* generated go code + hack generating scripts
* everything else

File: pkg/kubelet/apis/podresources/v1alpha1/api.proto (hunk 5 above)
Comment (dashpole) [reply pkg/kubelet/apis/podresources/v1alpha1/api.proto]: There was, nice catch.

File: cmd/kubelet/app/server.go (hunk 7 above)
Comment (dashpole) [reply cmd/kubelet/app/server.go]: Hmmm... it is an opt-in feature by nature of it being a new API.  @vishh WDYT?

File: pkg/kubelet/kubelet.go (hunk 6 above)
Comment (dashpole) [reply pkg/kubelet/kubelet.go]: I think I prefer keeping it with the creation of all of the other directories the kubelet makes.

File: cmd/kubelet/app/server.go (hunk 7 above)
Comment (dims) [reply cmd/kubelet/app/server.go]: +1 to a feature gate

Comment (dashpole) [reply cmd/kubelet/app/server.go]: added

File: hack/update-generated-pod-resources.sh [hunk 8]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
```
Comment (RenaudWasTaken) [inline hack/update-generated-pod-resources.sh]: You're missing `hack/validate-generated-pod-resources.sh`

File: pkg/features/kube_features.go [hunk 9]
```diff
@@ -378,6 +378,12 @@ const (
 	//
 	// Allow TTL controller to clean up Pods and Jobs after they finish.
 	TTLAfterFinished utilfeature.Feature = "TTLAfterFinished"
```
Comment (RenaudWasTaken) [inline pkg/features/kube_features.go]: Nit split this specific file in it's own commit :)

File: pkg/kubelet/apis/podresources/server.go [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+// podResourcesServer implements PodResourcesListerServer
+type podResourcesServer struct {
+	podsProvider    PodsProvider
+	devicesProvider DevicesProvider
+}
+
+// NewPodResourcesServer returns a PodResourcesListerServer which lists pods provided by the PodsProvider
+// with device information provided by the DevicesProvider
+func NewPodResourcesServer(podsProvider PodsProvider, devicesProvider DevicesProvider) v1alpha1.PodResourcesListerServer {
+	return &podResourcesServer{podsProvider: podsProvider, devicesProvider: devicesProvider}
```
Comment (RenaudWasTaken) [inline pkg/kubelet/apis/podresources/server.go]: Nit put initialization over multiple lines

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 11]
```diff
@@ -802,3 +803,8 @@ func (m *ManagerImpl) isDevicePluginResource(resource string) bool {
 	}
 	return false
 }
+
+// GetDevices returns the devices used by the specified container
+func (m *ManagerImpl) GetDevices(podUID, containerName string) []*podresourcesapi.ContainerDevices {
```
Comment (RenaudWasTaken) [inline pkg/kubelet/cm/devicemanager/manager.go]: This is the kind of issue that a mutex at the podDevices struct level would have prevented.
Code becomes pretty hard to maintain if it's not carefully reviewed with the mutex decision to make it "global".

You'll have to call `m.mutex.Lock()` and `m.mutex.Unlock()` here

**Review by RenaudWasTaken — COMMENTED** [review]
Almost good, I can lgtm after you've addressed my comments

File: hack/update-generated-pod-resources.sh (hunk 8 above)
Comment (dashpole) [reply hack/update-generated-pod-resources.sh]: added

File: pkg/features/kube_features.go (hunk 9 above)
Comment (dashpole) [reply pkg/features/kube_features.go]: done.

File: pkg/kubelet/apis/podresources/server.go (hunk 10 above)
Comment (dashpole) [reply pkg/kubelet/apis/podresources/server.go]: done.

File: pkg/kubelet/cm/devicemanager/manager.go (hunk 11 above)
Comment (dashpole) [reply pkg/kubelet/cm/devicemanager/manager.go]: good catch, done.

File: test/e2e_node/util.go [hunk 12]
```diff
@@ -62,6 +66,10 @@ var busyboxImage = imageutils.GetE2EImage(imageutils.BusyBox)
 const (
 	// Kubelet internal cgroup name for node allocatable cgroup.
 	defaultNodeAllocatableCgroup = "kubepods"
+	// defaultPodResourcesPath is the path to the local endpoint serving the podresources GRPC service.
+	defaultPodResourcesPath    = "/var/lib/kubelet/pod-resources"
+	defaultPodResourcesTimeout = 10 * time.Second
+	defaultPodResourcesMaxSize = 1024 * 1024 * 16 // 16 Mb
```
Comment (jiayingz) [inline test/e2e_node/util.go]: Do we really need this big buffer size?

**Review by jiayingz — APPROVED** [review]
/lgtm

File: pkg/kubelet/apis/podresources/v1alpha1/api.proto [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+// ListPodResourcesResponse is the response returned by List function
+message ListPodResourcesResponse {
+    repeated PodResources pod_resources = 1;
+}
+
+// PodResources contains information about the node resources assigned to a pod
+message PodResources {
+    string name = 1;
+    string namespace = 2;
```
Comment (vishh) [inline pkg/kubelet/apis/podresources/v1alpha1/api.proto]: How about pod UID?

Comment (dashpole) [reply pkg/kubelet/apis/podresources/v1alpha1/api.proto]: That isn't required by the use-cases covered in the KEP.  We can always add it in the future if that changes

Comment (dashpole) [reply pkg/kubelet/apis/podresources/v1alpha1/api.proto]: To further clarify, UIDs shouldn't be included in metrics based on the [kubernetes instrumentation guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/instrumentation.md#resource-referencing).  If we come across a use-case in the future where we need to get container metadata from the pod UID, then we can consider adding it.

**Comment by dashpole** [comment]
Consumers will definitely have to poll at some frequency, otherwise they would miss the addition of new pods, and keep monitoring devices long after they have stopped being used.

IIUC, you are suggesting consumers of this API will precompute the device associated with a particular pod, and avoid precomputing devices for pods they have already seen.  I think in most cases, precomputation would actually go the opposite direction, and consumers of this API will precompute the metadata for a particular device, which would not be susceptible to the issue you describe.  Regardless, I think it is a premature optimization at this point.

**Comment by vishh** [comment]
I forgot that this API is for reverse lookup (device -> pod). Thanks for
clarifying (again).
/approve

**Review by RenaudWasTaken — APPROVED** [review]
/lgtm
